test/python/test_composite.py

```diff
@@ -186,62 +186,62 @@ def test_empty_equilibrate(self):
             self.gas.equilibrate("TP")
 
 
-class TestEmptySolution(TestEmptyThermoPhase):
+class TestEmptySolution():
     """ Test empty Solution object """
-    @classmethod
-    def setUpClass(cls):
-        utilities.CanteraTest.setUpClass()
-        cls.gas = ct.Solution()
 
-    def test_empty_composite(self):
-        self.assertEqual(self.gas.thermo_model, "none")
-        self.assertEqual(self.gas.composite, ("none", "none", "none"))
+    @pytest.fixture(scope='function')
+    def gas(self):
+        return ct.Solution()
+
+    def test_empty_composite(self, gas):
+        assert gas.thermo_model == "none"
+        assert gas.composite ==  ("none", "none", "none")
 
 
-class TestEmptyEdgeCases(utilities.CanteraTest):
+class TestEmptyEdgeCases():
     """ Test for edge cases where constructors are not allowed """
     def test_empty_phase(self):
-        with self.assertRaisesRegex(ValueError, "Arguments are insufficient to define a phase"):
+        with pytest.raises(ValueError, match="Arguments are insufficient to define a phase"):
             ct.ThermoPhase(thermo="ideal-gas")
 
     def test_empty_kinetics(self):
-        with self.assertRaisesRegex(ValueError, "Cannot instantiate"):
+        with pytest.raises(ValueError, match="Cannot instantiate"):
             ct.Kinetics()
 
     def test_empty_transport(self):
-        with self.assertRaisesRegex(ValueError, "Cannot instantiate"):
+        with pytest.raises(ValueError, match="Cannot instantiate"):
             ct.Transport()
 
 
-class TestSolutionArray(utilities.CanteraTest):
+class TestSolutionArray():
     """ Test SolutionArray basics """
-    @classmethod
-    def setUpClass(cls):
-        utilities.CanteraTest.setUpClass()
-        cls.gas = ct.Solution('h2o2.yaml', transport_model=None)
-
-    def test_from_state_scalar(self):
-        state = list(self.gas.state)
-        arr = ct.SolutionArray(self.gas, states=[state])
+
+    @pytest.fixture(scope='function')
+    def gas(self):
+        return ct.Solution('h2o2.yaml', transport_model=None)
```

Comment:
I suppose one advantage of a new instance per test is that we're sure there aren't any state changes that are retained between tests and it ensures separate objects when using pytest-xdist.

If the performance penalty is too big, we could consider making this class or module scoped, rather than function scoped.

Reply:
There are presently 2,339 distinct test cases in the Python test suite, most of which require at least one `Solution` object. On my computer, the full test suite completes (in single threaded mode) in about 50 seconds. The time to instantiate a `Solution` object ranges from about 0.8 ms for `h2o2.yaml` without transport properties to 34 ms for `gri30.yaml` with transport properties. So in the best case, this represents ~4% of the total test time, which is small but not completely negligible. The case where it's perhaps worth optimizing this is for those tests where we need to use larger mechanisms or need transport properties, as the instantiation of those objects is a bit slow (and of course this all slows down more if you need to use debug builds).

Currently, we do often re-use `Solution` objects within a test class, though skimming the test code I see there are more cases when we re-instantiate the `Solution` for every test than I thought. Resetting `TPY` at the start of each test should generally be sufficient to avoid issues, with the exception of those tests that add/modify species and reactions.

The parallel testing using `pytest-xdist` isn't an issue, as that creates separate processes that can't share objects in any case.

Reply:
I was a bit wary of setting the gas at the class level because of tests like the one below that alters the state of the shared object between all the tests.

```
 def test_slice_twice(self):
        T_list = np.linspace(300, 1000, 8)
        self.gas.TPX = T_list[0], ct.one_atm, {"H2": 1.}
        arr = ct.SolutionArray(self.gas)
        for T in T_list[1:]:
            self.gas.TPX = T, ct.one_atm, {"H2": 1.}
            arr.append(self.gas.state)
        ix = 4
        arr_trunc = arr[ix:]
        assert arr_trunc.T[0] == arr.T[ix]
        assert arr_trunc[0].T == arr.T[ix]
        assert arr_trunc.T[-1] == arr.T[-1]
        assert arr_trunc[-1].T == arr.T[-1]
        assert (arr_trunc.T[:2] == arr.T[ix:ix+2]).all()
        assert (arr_trunc[:2].T == arr.T[ix:ix+2]).all()
        with self.assertRaises(IndexError):
            arr_trunc[10]
            

```